### PR TITLE
feat(capp): extensible event source framework

### PR DESCRIPTION
**Pull request by dvirgilad**
## Summary

- Introduces a pluggable Knative Eventing source system — new source types implement `EventSourceKind` and self-register; the controller dispatches through a registry without knowing concrete types
- Adds `EventSourcesSpec` to `CappSpec` with a required, unique `Name` per source; webhook enforces uniqueness, CRD schema enforces presence
- Adds `EventingStatus` to `CappStatus` aggregated by `EventSourceManager.GetStatus` across all registered managers
- Adds generic `eventSourceWatchPredicate[T]` in the controller covering spec and readiness transitions
- Adds `docs/adding-event-sources.md`: 8-step guide covering API type, interface impl, dispatch wiring, blank import, RBAC annotation, Watch + predicate, tests, and CRD sync

## Test plan

- [ ] Unit tests pass: `go test ./internal/kinds/capp/sources/... ./internal/kinds/capp/resourcemanagers/...`
- [ ] `IsRequired` returns false for empty sources, true otherwise
- [ ] `Manage` routes to `createOrUpdateSources` when sources present, `CleanUp` when not
- [ ] `GetStatus` aggregates across all registered managers and propagates errors
- [ ] Webhook rejects duplicate source names
- [ ] CRD schema rejects missing or empty `name` field
- [ ] `make manifests` produces CRDs with `required: [name]`, `minLength: 1`, and `eventingStatus` in status

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: api/v1alpha1/capp_types.go [hunk 1]
```diff
@@ -51,6 +51,50 @@ type CappSpec struct {
...
 	// VolumesSpec defines the volumes specification for the Capp.
 	VolumesSpec VolumesSpec `json:"volumesSpec,omitempty"`
+
+	// EventSourcesSpec defines the event sources for the Capp.
+	// +optional
+	EventSourcesSpec EventSourcesSpec `json:"eventSourcesSpec,omitempty"`
+}
+
+// EventSource defines a single Knative Eventing source connected to the Capp.
+// Add source-type fields (e.g., PingSource *PingSourceSpec) as new source types are implemented.
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go]: Drop or relocate the “Add source-type fields…” sentence — it’s maintainer guidance, not API-user godoc.

File: api/v1alpha1/capp_types.go [hunk 2]
```diff
@@ -51,6 +51,50 @@ type CappSpec struct {
...
+type EventSource struct {
+	// Name is a unique logical identifier for this source within the Capp.
+	// Used to construct the underlying Kubernetes resource name as <capp-name>-<name>.
+	// Must be unique within the Capp's EventSourcesSpec.Sources list.
+	// +kubebuilder:validation:Required
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go]: Consider one tight sentence.

File: internal/kinds/capp/controllers/controller.go [hunk 3]
```diff
@@ -115,6 +115,7 @@ func (r *CappReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			handler.EnqueueRequestsFromMapFunc(r.findCappFromEvent),
 			builder.WithPredicates(predicate.GenerationChangedPredicate{}),
 		).
+		// Add one Watches block per event source type. See docs/adding-event-sources.md step 6.
```
Comment (sbahar619) [inline internal/kinds/capp/controllers/controller.go]: Redundant.

File: api/v1alpha1/capp_types.go [hunk 4]
```diff
@@ -51,6 +51,48 @@ type CappSpec struct {
 
 	// VolumesSpec defines the volumes specification for the Capp.
 	VolumesSpec VolumesSpec `json:"volumesSpec,omitempty"`
+
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go:54]: The API types placement breaks the natural “spec types together / status types together” flow—consider regrouping so the file reads in one coherent order without changing the API.

File: api/v1alpha1/capp_types.go [hunk 5]
```diff
@@ -51,6 +51,48 @@ type CappSpec struct {
...
+	// EventSourcesSpec defines the event sources for the Capp.
+	// +optional
+	EventSourcesSpec EventSourcesSpec `json:"eventSourcesSpec,omitempty"`
+}
+
+// EventSource defines a single Knative Eventing source connected to the Capp.
+type EventSource struct {
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go]: EventSource overlaps and is easy to confuse with Knative. Either add GoDoc or rename to something scoped.

File: api/v1alpha1/capp_types.go [hunk 6]
```diff
@@ -51,6 +51,48 @@ type CappSpec struct {
...
+
+// EventSourceStatus shows the observed state of a single event source resource.
+type EventSourceStatus struct {
+	// Name is the K8s name of the underlying source resource.
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// Ready indicates whether the underlying source resource is ready.
+	// +optional
+	Ready bool `json:"ready,omitempty"`
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go]: omitempty on ready hides false in JSON. Prefer the standard metav1.Condition shape, so “not ready” is always visible and consistent with the rest of Kubernetes APIs.

File: internal/kinds/capp/resourcemanagers/eventsource.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+// It lists owned resources per manager so it handles the case where all sources were removed from spec.
+func (e EventSourceManager) CleanUp(capp cappv1alpha1.Capp) error {
+	rm := rclient.ResourceManagerClient{Ctx: e.Ctx, K8sclient: e.K8sclient, Log: e.Log}
+	for _, manager := range sources.AllManagers() {
+		owned, err := manager.List(e.Ctx, rm, e.Log, capp)
+		if err != nil {
+			return fmt.Errorf("failed to list owned event sources for cleanup: %w", err)
+		}
+		for _, obj := range owned {
+			if err := e.K8sclient.Delete(e.Ctx, obj); err != nil {
```
Comment (sbahar619) [inline internal/kinds/capp/resourcemanagers/eventsource.go]: Use client.IgnoreNotFound around Delete, consistent with other managers.

**Review by sbahar619 — COMMENTED** [review]
Thanks, Looks good, I left few comments, I will continue review in small cycles.

File: internal/kinds/capp/sources/sources.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+var (
+	registryMu              sync.RWMutex
+	eventSourceKindRegistry = make(map[string]EventSourceKind)
+)
+
+func init() {
+	// Register source managers here as new source types are added.
+}
```
Comment (sbahar619) [inline internal/kinds/capp/sources/sources.go]: Redundant: this init() does nothing. Remove it; add init() again in a later PR only when it actually registers source managers.

File: internal/kinds/capp/sources/sources.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+// Register adds an EventSourceKind implementation under the given source type key.
+// Pass nil to remove a previously registered manager (useful in tests).
+// Call this from an init() function in the source-type package.
```
Comment (sbahar619) [inline internal/kinds/capp/sources/sources.go]: Register’s GoDoc reads more like an implementation guide than an API summary—tighten it to the public contract.

File: internal/kinds/capp/sources/sources.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	defer registryMu.Unlock()
+	if manager == nil {
+		delete(eventSourceKindRegistry, sourceType)
+		return
+	}
+	eventSourceKindRegistry[sourceType] = manager
+}
+
+// AllManagers returns all registered EventSourceKind implementations.
+func AllManagers() []EventSourceKind {
```
Comment (sbahar619) [inline internal/kinds/capp/sources/sources.go]: AllManagers is misleading—it returns []EventSourceKind. Consider RegisteredEventSourceKinds for accuracy.

File: internal/kinds/capp/sources/registry.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
+package sources
```
Comment (sbahar619) [inline internal/kinds/capp/sources/registry.go:1]: Consider renaming sources.go to registry.go to better match the file’s role, if you want the layout to read more clearly.

File: internal/kinds/capp/sources/sources.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	for _, m := range eventSourceKindRegistry {
+		result = append(result, m)
+	}
+	return result
+}
+
+// GetEventSourceKindManager returns the registered manager for the source type encoded in source.
+// Returns false when no source-type field is set or the type has no registered manager.
+// Extend the switch below when adding a new source type field to EventSource.
+func GetEventSourceKindManager(source cappv1alpha1.EventSource) (EventSourceKind, bool) {
```
Comment (sbahar619) [inline internal/kinds/capp/sources/sources.go]: GetEventSourceKindManager returns an EventSourceKind, but the name says “Manager”—consider renaming GetEventSourceKind so the symbol matches the type it returns.

File: internal/kinds/capp/sources/sources.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	sourceType := ""
+	switch {
+	// Add a case for each new source-type field added to EventSource, e.g.:
+	// case source.PingSource != nil:
+	//     sourceType = eventsources.PingSourceType
+	}
```
Comment (sbahar619) [inline internal/kinds/capp/sources/sources.go]: The empty switch {} is redundant until real cases exist, reintroduce the switch when really used.

File: internal/kinds/capp/sources/sources_test.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	assert.True(t, found, "registered manager should appear in AllManagers")
+}
+
+func TestAllManagers_empty(t *testing.T) {
+	// Registry may have entries from other tests, but AllManagers must not panic.
+	require.NotPanics(t, func() { sources.AllManagers() })
+}
+
+func TestStubEventSourceKind_GetStatus_returnsError(t *testing.T) {
```
Comment (sbahar619) [inline internal/kinds/capp/sources/sources_test.go]: Please align test names with Test<FuncName> (no extra type/prefix in the outer name; put behavior in t.Run(...)).

Example: rename TestStubEventSourceKind_GetStatus_returnsError → TestGetStatus, with a subtest like t.Run("returns error from backend", ...) if you need the scenario spelled out.

File: docs/contributing/adding-event-sources.md [hunk 15]
```diff
@@ -0,0 +1,188 @@
+# Adding a New Event Source Type
```
Comment (sbahar619) [inline docs/contributing/adding-event-sources.md:1]: Feels like implementation/contributor content, not general docs/ material—redundant with what the first real source PR will demonstrate. If you keep it, move it under something like docs/implementation/ (or docs/contributing/) instead of the docs root.

File: api/v1alpha1/capp_types.go [hunk 16]
```diff
@@ -51,6 +51,25 @@ type CappSpec struct {
...
+	// +optional
+	Sources []SourceConfiguration `json:"sources,omitempty"`
+}
+
+// SourceConfiguration defines a single Knative Eventing source connected to the Capp.
+type SourceConfiguration struct {
+	// Name is a unique logical identifier for this source within the Capp.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go]: Still hard to scan: event-source helper types sit between CappSpec and ScaleSpec. Consider moving EventSourcesSpec + SourceConfiguration next to other spec “leaf” types (or after ScaleSpec) so spec structs read top-down.

File: api/v1alpha1/capp_types.go [hunk 17]
```diff
@@ -182,6 +201,28 @@ type StateStatus struct {
...
+
+// EventSourceStatus shows the observed state of a single event source resource.
+type EventSourceStatus struct {
+	// Name is the K8s name of the underlying source resource.
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// Ready indicates whether the underlying source resource is ready.
+	// +optional
+	Ready metav1.ConditionStatus `json:"ready,omitempty"`
```
Comment (sbahar619) [inline api/v1alpha1/capp_types.go]: We moved to ConditionStatus (good), but the earlier ask was the full metav1.Condition shape for stable semantics/reasons. OK to keep this if intentional.

File: internal/kinds/capp/resourcemanagers/eventsource_test.go [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.Equal(t, tt.expected, newManager().IsRequired(tt.capp))
+		})
+	}
+}
+
+func TestEventSourceManager_Manage_callsCreateOrUpdate_when_sources_present(t *testing.T) {
```
Comment (sbahar619) [inline internal/kinds/capp/resourcemanagers/eventsource_test.go]: Outer test names still use TestEventSourceManager_*; I suggest to use convention of Test<FuncName> + t.Run for scenarios.

File: internal/kinds/capp/sources/registry.go [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+	cappv1alpha1 "github.com/dana-team/container-app-operator/api/v1alpha1"
+	rclient "github.com/dana-team/container-app-operator/internal/kinds/capp/resourceclient"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// EventSourceManager defines the interface for managing a specific Knative Eventing source type.
+// Each source type (e.g., PingSource, ApiServerSource) should implement this interface to handle its lifecycle.
+type EventSourceManager interface {
```
Comment (sbahar619) [inline internal/kinds/capp/sources/registry.go]: We now have two different EventSourceManager symbols (sources interface vs resourcemanagers struct), which is easy to misread in reviews/imports. AllManagers / GetEventSourceKindManager also reuse “manager” in a way that doesn’t match what they return. Please rename for disambiguation (interface + registry helpers, and/or the coordinator struct) so “manager” means one thing.

**Review by sbahar619 — COMMENTED** [review]
Thanks!
Looks good!
I added few comments...

File: internal/kinds/capp/resourcemanagers/eventsource.go [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+func (e EventSourceManager) Manage(capp cappv1alpha1.Capp) error {
+	if e.IsRequired(capp) {
+		return e.createOrUpdateSources(capp)
+	}
+	return e.CleanUp(capp)
+}
+
+// CleanUp deletes all source resources owned by the Capp, across all registered source types.
+// It lists owned resources per manager so it handles the case where all sources were removed from spec.
+func (e EventSourceManager) CleanUp(capp cappv1alpha1.Capp) error {
```
Comment (sbahar619) [inline internal/kinds/capp/resourcemanagers/eventsource.go:44]: Please set controller owner refs on created event-source objects (same idea as KnativeServiceManager) so GC/cleanup and ownership checks behave consistently.

File: internal/kinds/capp/sources/registry_test.go [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, tc.run)
+	}
+}
+
+func TestGetStatus(t *testing.T) {
```
Comment (sbahar619) [inline internal/kinds/capp/sources/registry_test.go]: TestGetStatus calls the test stub’s GetStatus directly, so it never exercises sources. The “status error propagates” behavior is already covered by TestEventSourceManager_GetStatus_propagates_error in eventsource_test.go .

File: internal/kinds/capp/sources/registry_test.go [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+						break
+					}
+				}
+				assert.True(t, found, "registered manager should appear in AllManagers")
+			},
+		},
+		{
+			name: "all managers does not panic",
+			run: func(t *testing.T) {
+				require.NotPanics(t, func() { sources.AllManagers() })
```
Comment (sbahar619) [inline internal/kinds/capp/sources/registry_test.go]: This subtest only checks NotPanics(AllManagers), which doesn’t encode a meaningful contract for an empty registry. Drop it unless there’s a specific regression you’re guarding against.

File: internal/webhook/rcs/v1alpha1/validator.go [hunk 23]
```diff
@@ -126,3 +130,14 @@ func validateNFSVolumeMounts(capp cappv1alpha1.Capp) error {
 
 	return fmt.Errorf("invalid nfsVolumes: volumes [%s] must be mounted by at least one container", strings.Join(missingVolumeNames, ", "))
 }
+
+func validateEventSources(capp cappv1alpha1.Capp) error {
```
Comment (sbahar619) [inline internal/webhook/rcs/v1alpha1/validator.go:134]: Please add a validator unit test that rejects duplicate eventSourcesSpec.sources[].name values, so the webhook rule can’t regress silently.

File: docs/contributing/adding-event-sources.md (hunk 15 above)
Comment (dvirgilad) [reply docs/contributing/adding-event-sources.md:1]: Since it is non-trivial and looking at commits is not always the most straight-forward, i added this. Moved to docs/contributing

**Review by sbahar619 — APPROVED** [review]
/lgtm
